please patch the CQ:
```text
Run actions/cache@v4
Cache hit for: pre-commit-3|/opt/hostedtoolcache/Python/3.13.7/x64|1a9a586caed026458236911727580546166b820cd4255a17ea61253126733d6c
Received 48289771 of 48289771 (100.0%), 133.1 MBs/sec
Cache Size: ~46 MB (48289771 B)
/usr/bin/tar -xf /home/runner/work/_temp/b1be561c-9f3e-4e49-a236-2eeec0b71a32/cache.tzst -P -C /home/runner/work/LMCache/LMCache --use-compress-program unzstd
Cache restored successfully
Cache restored from key: pre-commit-3|/opt/hostedtoolcache/Python/3.13.7/x64|1a9a586caed026458236911727580546166b820cd4255a17ea61253126733d6c
Run pre-commit run --show-diff-on-failure --color=always --all-files
Check SPDX headers.......................................................Passed
isort....................................................................Passed
ruff.....................................................................Failed
- hook id: ruff
- exit code: 1
- files were modified by this hook

Error: lmcache/v1/storage_backend/connector/redis_connector.py:517:89: E501 Line too long (101 > 88)

ruff-format..............................................................Failed
- hook id: ruff-format
- files were modified by this hook

1 file reformatted, 242 files left unchanged

codespell................................................................Passed
clang-format.............................................................Passed
mypy.....................................................................Passed
pre-commit hook(s) made changes.
If you are seeing this message in CI, reproduce locally with: `pre-commit run --all-files`.
To run `pre-commit` as part of git workflow, use `pre-commit install`.
All changes made by hooks:
diff --git a/lmcache/v1/storage_backend/connector/redis_connector.py b/lmcache/v1/storage_backend/connector/redis_connector.py
index 7f27255..38f3b82 100644
--- a/lmcache/v1/storage_backend/connector/redis_connector.py
+++ b/lmcache/v1/storage_backend/connector/redis_connector.py
@@ -514,7 +514,10 @@ class RedisClusterConnector(RemoteConnector):
     ):
         # calling self.put will create a circular dependency
         await asyncio.gather(
-            *(self._put(key, memory_obj) for key, memory_obj in zip(keys, memory_objs))
+            *(
+                self._put(key, memory_obj)
+                for key, memory_obj in zip(keys, memory_objs, strict=False)
+            )
         )
 
     async def batched_put(
Error: Process completed with exit code 1.
```

--- a/lmcache/v1/storage_backend/connector/redis_adapter.py
+++ b/lmcache/v1/storage_backend/connector/redis_adapter.py
@@ -75,3 +75,48 @@ class RedisSentinelConnectorAdapter(ConnectorAdapter):
             loop=context.loop,
             local_cpu_backend=context.local_cpu_backend,
         )
+
+
+class RedisClusterConnectorAdapter(ConnectorAdapter):
+    """Adapter for Redis Cluster connectors."""
+
+    def __init__(self) -> None:
+        super().__init__("redis-cluster://")
+
+    def can_parse(self, url: str) -> bool:
+        return url.startswith(self.schema)
+
+    def create_connector(self, context: ConnectorContext) -> RemoteConnector:
+        # Local
+        from .redis_connector import RedisClusterConnector
+
+        logger.info(f"Creating Redis Cluster connector for URL: {context.url}")
+        url = context.url[len(self.schema) :]
+
+        # Parse username and password
+        username: str = ""
+        password: str = ""
+        if "@" in url:
+            auth, url = url.split("@", 1)
+            if ":" in auth:
+                username, password = auth.split(":", 1)
+            else:
+                username = auth
+
+        # Parse host and port
+        hosts_and_ports: List[Tuple[str, int]] = []
+        assert self.schema is not None
+        for sub_url in url.split(","):
+            if not sub_url.startswith(self.schema):
+                sub_url = self.schema + sub_url
+
+            parsed_url = parse_remote_url(sub_url)
+            hosts_and_ports.append((parsed_url.host, parsed_url.port))
+
+        return RedisClusterConnector(
+            hosts_and_ports=hosts_and_ports,
+            username=username,
+            password=password,
+            loop=context.loop,
+            local_cpu_backend=context.local_cpu_backend,
+        )

--- a/lmcache/v1/storage_backend/connector/redis_connector.py
+++ b/lmcache/v1/storage_backend/connector/redis_connector.py
@@ -7,6 +7,7 @@ import inspect
 import os
 
 # Third Party
+from redis.asyncio.cluster import ClusterNode, RedisCluster
 import redis.asyncio as redis
 
 # First Party
@@ -384,3 +385,229 @@ class RedisSentinelConnector(RemoteConnector):
     async def close(self):
         self.master.close()
         self.slave.close()
+
+
+class RedisClusterConnector(RemoteConnector):
+    """
+    The remote url starts with "redis-cluster:// and can include one or
+    multiple hosts:ports, separated by commas.
+
+    Example:
+        remote_url: "redis-cluster://host1:7000,host2:7000,host3:7000"
+
+    Extra environment variables:
+    - REDIS_TIMEOUT (optional) -- Timeout in seconds, default is 1 if not set
+    """
+
+    def __init__(
+        self,
+        hosts_and_ports: List[Tuple[str, int]],
+        username: str,
+        password: str,
+        loop: asyncio.AbstractEventLoop,
+        local_cpu_backend: LocalCPUBackend,
+    ):
+        # Convert hosts_and_ports to startup_nodes format expected by RedisCluster
+        startup_nodes = [ClusterNode(h, p) for (h, p) in hosts_and_ports]
+
+        # set a large max
+        self.max_connections = 150
+        # redis will crash if we have more than max_connections connections
+        self.sem = asyncio.Semaphore(self.max_connections)
+
+        # Initialize cluster connection
+        self.cluster = RedisCluster(
+            startup_nodes=startup_nodes,
+            username=username,
+            password=password,
+            max_connections=self.max_connections,
+            decode_responses=False,
+        )
+        self.loop = loop
+        self.local_cpu_backend = local_cpu_backend
+
+        self.pq_executor = AsyncPQExecutor(loop)
+
+    async def _exists(self, key: CacheEngineKey) -> bool:
+        async with self.sem:
+            return bool(await self.cluster.exists(key.to_string() + "metadata"))
+
+    async def exists(self, key: CacheEngineKey) -> bool:
+        return await self.pq_executor.submit_job(
+            self._exists, key=key, priority=Priorities.PEEK
+        )
+
+    def exists_sync(self, key: CacheEngineKey) -> bool:
+        future = asyncio.run_coroutine_threadsafe(self.exists(key), self.loop)
+        return bool(future.result())
+
+    async def _get(self, key: CacheEngineKey) -> Optional[MemoryObj]:
+        key_str = key.to_string()
+        async with self.sem:
+            metadata_bytes = await self.cluster.get(key_str + "metadata")
+
+            if metadata_bytes is None:
+                return None
+
+            assert not inspect.isawaitable(metadata_bytes)
+
+            metadata = RemoteMetadata.deserialize(memoryview(metadata_bytes))
+
+            memory_obj = self.local_cpu_backend.allocate(
+                metadata.shape,
+                metadata.dtype,
+                metadata.fmt,
+            )
+            if memory_obj is None:
+                logger.warning("Failed to allocate memory during remote receive")
+                return None
+
+            # TODO(Jiayi): Find a way to do `get` inplace
+            kv_bytes = await self.cluster.get(key_str + "kv_bytes")
+
+        assert not inspect.isawaitable(kv_bytes)
+
+        if kv_bytes is None:
+            # TODO (Jiayi): We might need a way to better handle
+            # consistency issues.
+            # TODO (Jiayi): A better way is to aggregate metadata
+            # and kv cache in one key.
+            logger.warning(
+                "Key exists but KV cache does not exist."
+                "Might happen when the cache is evicted by redis."
+            )
+            async with self.sem:
+                await self.cluster.delete(key_str + "metadata")
+            return None
+
+        if isinstance(memory_obj.byte_array, memoryview):
+            view = memory_obj.byte_array
+            if view.format == "<B":
+                view = view.cast("B")
+        else:
+            view = memoryview(memory_obj.byte_array)
+
+        if isinstance(kv_bytes, (bytes, bytearray)):
+            view[: metadata.length] = kv_bytes
+        elif isinstance(kv_bytes, str):
+            converted = kv_bytes.encode("utf-8")
+            view[: metadata.length] = converted
+        else:
+            converted = bytes(kv_bytes)
+            view[: metadata.length] = converted
+
+        return memory_obj
+
+    async def get(self, key: CacheEngineKey) -> Optional[MemoryObj]:
+        return await self.pq_executor.submit_job(
+            self._get, key=key, priority=Priorities.GET
+        )
+
+    def support_batched_put(self) -> bool:
+        return True
+
+    async def _batched_put(
+        self, keys: List[CacheEngineKey], memory_objs: List[MemoryObj]
+    ):
+        # calling self.put will create a circular dependency
+        await asyncio.gather(
+            *(self._put(key, memory_obj) for key, memory_obj in zip(keys, memory_objs))
+        )
+
+    async def batched_put(
+        self, keys: List[CacheEngineKey], memory_objs: List[MemoryObj]
+    ):
+        await self.pq_executor.submit_job(
+            self._batched_put,
+            keys=keys,
+            memory_objs=memory_objs,
+            priority=Priorities.PUT,
+        )
+
+    async def _put(self, key: CacheEngineKey, memory_obj: MemoryObj):
+        # TODO(Jiayi): The following code is ugly.
+        # Please use a function like `memory_obj.to_meta()`.
+        kv_bytes = memory_obj.byte_array
+        kv_shape = memory_obj.get_shape()
+        kv_dtype = memory_obj.get_dtype()
+        memory_format = memory_obj.get_memory_format()
+
+        metadata_bytes = RemoteMetadata(
+            len(kv_bytes), kv_shape, kv_dtype, memory_format
+        ).serialize()
+
+        key_str = key.to_string()
+        # kv bytes needs to be set first to avoid race condition
+        async with self.sem:
+            await self.cluster.set(key_str + "kv_bytes", kv_bytes)
+            await self.cluster.set(key_str + "metadata", metadata_bytes)
+
+    async def put(self, key: CacheEngineKey, memory_obj: MemoryObj):
+        await self.pq_executor.submit_job(
+            self._put, key=key, memory_obj=memory_obj, priority=Priorities.PUT
+        )
+
+    # TODO
+    @no_type_check
+    async def list(self) -> List[str]:
+        pass
+
+    async def close(self):
+        await self.pq_executor.shutdown(wait=True)
+        await self.cluster.close()
+        logger.info("Closed the redis cluster connection")
+
+    def support_batched_async_contains(self) -> bool:
+        return True
+
+    async def _batched_async_contains(
+        self,
+        lookup_id: str,
+        keys: List[CacheEngineKey],
+        pin: bool = False,
+    ) -> int:
+        num_hit_counts = 0
+        for key in keys:
+            async with self.sem:
+                if not await self.cluster.exists(key.to_string() + "metadata"):
+                    return num_hit_counts
+            num_hit_counts += 1
+        return num_hit_counts
+
+    async def batched_async_contains(
+        self,
+        lookup_id: str,
+        keys: List[CacheEngineKey],
+        pin: bool = False,
+    ) -> int:
+        return await self.pq_executor.submit_job(
+            self._batched_async_contains,
+            lookup_id=lookup_id,
+            keys=keys,
+            pin=pin,
+            priority=Priorities.PEEK,
+        )
+
+    def support_batched_get_non_blocking(self) -> bool:
+        return True
+
+    async def _batched_get_non_blocking(
+        self,
+        lookup_id: str,
+        keys: List[CacheEngineKey],
+    ) -> List[MemoryObj]:
+        # calling self.get will create a circular dependency
+        results = await asyncio.gather(*(self._get(key) for key in keys))
+        return [r for r in results if r is not None]
+
+    async def batched_get_non_blocking(
+        self,
+        lookup_id: str,
+        keys: List[CacheEngineKey],
+    ) -> List[MemoryObj]:
+        return await self.pq_executor.submit_job(
+            self._batched_get_non_blocking,
+            lookup_id=lookup_id,
+            keys=keys,
+            priority=Priorities.PREFETCH,
+        )

--- a/lmcache/v1/storage_backend/connector/redis_connector.py
+++ b/lmcache/v1/storage_backend/connector/redis_connector.py
@@ -140,7 +140,10 @@ class RedisConnector(RemoteConnector):
     ):
         # calling self.put will create a circular dependency
         await asyncio.gather(
-            *(self._put(key, memory_obj) for key, memory_obj in zip(keys, memory_objs))
+            *(
+                self._put(key, memory_obj)
+                for key, memory_obj in zip(keys, memory_objs, strict=False)
+            )
         )
 
     async def batched_put(
@@ -511,7 +514,10 @@ class RedisClusterConnector(RemoteConnector):
     ):
         # calling self.put will create a circular dependency
         await asyncio.gather(
-            *(self._put(key, memory_obj) for key, memory_obj in zip(keys, memory_objs))
+            *(
+                self._put(key, memory_obj)
+                for key, memory_obj in zip(keys, memory_objs, strict=False)
+            )
         )
 
     async def batched_put(

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -221,6 +221,32 @@ class MockRedisSentinel:
         return self.slave_redis
 
 
+class MockRedisCluster:
+    def __init__(
+        self, startup_nodes=None, max_connections=None, decode_responses=False, **kwargs
+    ):
+        self.startup_nodes = startup_nodes or []
+        self.max_connections = max_connections
+        self.decode_responses = decode_responses
+        self.store = {}
+
+    async def set(self, key, value):
+        self.store[key] = value
+        return True
+
+    async def get(self, key):
+        return self.store.get(key, None)
+
+    async def exists(self, key):
+        return key in self.store
+
+    async def delete(self, key):
+        return self.store.pop(key, None) is not None
+
+    async def close(self):
+        pass
+
+
 @dataclass
 class LMCacheServerProcess:
     server_url: str
@@ -246,6 +272,12 @@ def mock_redis_sentinel():
         patch("redis.Sentinel", MockRedisSentinel) as mock,
         patch("redis.asyncio.Sentinel", MockRedisSentinel),
     ):
+        yield mock
+
+
+@pytest.fixture(scope="function", autouse=True)
+def mock_redis_cluster():
+    with patch("redis.asyncio.cluster.RedisCluster", MockRedisCluster) as mock:
         yield mock
 
 

--- a/tests/v1/test_connector.py
+++ b/tests/v1/test_connector.py
@@ -12,6 +12,7 @@ import torch
 from lmcache.config import LMCacheEngineMetadata
 from lmcache.v1.config import LMCacheEngineConfig
 from lmcache.v1.memory_management import PinMemoryAllocator
+from lmcache.v1.protocol import RemoteMetadata
 from lmcache.v1.storage_backend.connector import CreateConnector
 
 # Local
@@ -296,4 +297,104 @@ def test_redis_sentinel_connector(url, autorelease_v1):
 
     close_asyncio_loop(async_loop, async_thread)
 
+    memory_allocator.close()
+
+
+REDIS_CLUSTER_URLS = [
+    "redis-cluster://host1:7000,host2:7000,host3:7000",
+    "redis-cluster://clustercfg.cluster-name.id.region.cache.amazonaws.com:6379",
+    "redis-cluster://user:password@host1:7000,host2:7000,host3:7000",
+]
+
+
+@pytest.mark.parametrize("url", REDIS_CLUSTER_URLS)
+def test_redis_cluster_connector(url, autorelease_v1):
+    """Test Redis Cluster connector: exists, put, get operations.
+
+    This test uses the MockRedisCluster from conftest.py to simulate
+    Redis Cluster behavior without requiring an actual Redis Cluster setup.
+    """
+
+    # Standard
+    import os
+
+    os.environ["REDIS_TIMEOUT"] = "3.5"
+
+    async_loop, async_thread = init_asyncio_loop()
+    memory_allocator = PinMemoryAllocator(1024 * 1024 * 1024)
+
+    connector = autorelease_v1(CreateConnector(url, async_loop, memory_allocator))
+
+    random_key = dumb_cache_engine_key()
+
+    # Test 1: Verify key doesn't exist initially, test contains key not exist
+    future = asyncio.run_coroutine_threadsafe(connector.exists(random_key), async_loop)
+    assert not future.result()
+
+    # Test 2: Create and store test data
+    num_tokens = 1000
+    mem_obj_shape = [2, 32, num_tokens, 1024]
+    dtype = torch.bfloat16
+    memory_obj = memory_allocator.allocate(mem_obj_shape, dtype)
+    memory_obj.ref_count_up()
+
+    # Fill with deterministic test data
+    torch.manual_seed(42)
+    test_tensor = torch.randint(0, 100, memory_obj.raw_data.shape, dtype=torch.int64)
+    memory_obj.raw_data.copy_(test_tensor.to(torch.float32).to(dtype))
+
+    # Test 3: Put data
+    future = asyncio.run_coroutine_threadsafe(
+        connector.put(random_key, memory_obj), async_loop
+    )
+    future.result()
+
+    # Test 4: Verify key exists after putting data, test contains key exists
+    future = asyncio.run_coroutine_threadsafe(connector.exists(random_key), async_loop)
+    assert future.result()
+    assert memory_obj.get_ref_count() == 1
+
+    # Test 5: Retrieve and verify data
+    future = asyncio.run_coroutine_threadsafe(connector.get(random_key), async_loop)
+    retrieved_memory_obj = future.result()
+
+    check_mem_obj_equal([retrieved_memory_obj], [memory_obj])
+
+    close_asyncio_loop(async_loop, async_thread)
+    memory_allocator.close()
+
+
+@pytest.mark.parametrize("url", REDIS_CLUSTER_URLS)
+def test_cluster_metadata_without_kv_bytes(url, autorelease_v1):
+    async_loop, async_thread = init_asyncio_loop()
+    memory_allocator = PinMemoryAllocator(1024 * 1024 * 1024)
+    connector = autorelease_v1(CreateConnector(url, async_loop, memory_allocator))
+
+    random_key = dumb_cache_engine_key()
+    # build a small mem obj to get correct metadata bytes
+    memory_obj = memory_allocator.allocate([2, 32, 8, 64], torch.bfloat16)
+    kv_bytes = memory_obj.byte_array
+    meta = RemoteMetadata(
+        len(kv_bytes),
+        memory_obj.get_shape(),
+        memory_obj.get_dtype(),
+        memory_obj.get_memory_format(),
+    )
+    metadata_bytes = meta.serialize()
+
+    # clean up memory object after getting metadata
+    memory_obj.ref_count_down()
+
+    # inject only metadata, no kv_bytes
+    meta_key = random_key.to_string() + "metadata"
+    connector._connector.cluster.set(meta_key, metadata_bytes)
+
+    # get() should return None and remove the metadata without kv_bytes pair
+    future = asyncio.run_coroutine_threadsafe(connector.get(random_key), async_loop)
+    assert future.result() is None
+
+    future = asyncio.run_coroutine_threadsafe(connector.exists(random_key), async_loop)
+    assert not future.result()
+
+    close_asyncio_loop(async_loop, async_thread)
     memory_allocator.close()

